Fix alignment detection for long integers on 32-bits platforms

shindere opened the pull request:
This should fix Issue #7938.

xavierleroy commented on configure.ac:
```diff
@@ -698,14 +698,13 @@ AC_CHECK_ALIGNOF([double])
-)
+  [AS_CASE([$target_cpu],
+    [i686], [],
+    [AS_IF([test x"$ac_cv_alignof_double" = "x8" ],
```
Technically, you want to define `ARCH_ALIGN_DOUBLE` if the alignment of type `double` is greater than one word, namely 4 in the case of a 32-bit platform.  Imagine a weird platform where `double` must be 16-aligned...  To deal with this unlikely case, you may want to write
```
test "$ac_cv_alignof_double" -gt 4
```

xavierleroy commented on configure.ac:
```diff
@@ -698,14 +698,13 @@ AC_CHECK_ALIGNOF([double])
+    [i686], [],
+    [AS_IF([test x"$ac_cv_alignof_double" = "x8" ],
+      [AC_DEFINE([ARCH_ALIGN_DOUBLE], [1])])
+    AS_IF([test x"$ac_cv_alignof_long" = "x8" ],
```
Likewise, you may want to write
```
test "$ac_cv_alignof_long" -gt 4
```

xavierleroy reviewed (approved):
Looks good to me.  One suggested clarification below, which should not make any difference on the usual architectures.  Feel free to use or not.

shindere committed:
Fix alignment detection for long integers on 32-bits platforms

shindere commented:
Many thanks for your review, @xavierleroy.

I implemented your suggestion because I liked it a lot and rebased to
resolve the conflict in Changes.

Before that CI was happy so merging.

shindere commented:
Merged in trunk and cherry-picked to 4.08 as commit
6243c8787f5a8cab0b5bc6a1540ae5794de923a1
